File: Divide and Conquer Algos/Closest Pair Problem/code.py

# QUESTION
# Say there are n different ships moving in the sea. Write and implement an algorithm so
# that you can efficiently avoid collisions of the ship


# CODE : Divide and Conquer Algorithm
import math

def getDistance(p1, p2):
    y_sq = (p2[1]-p1[1])**2
    x_sq = (p2[0]-p1[0])**2
    return math.sqrt(y_sq + x_sq)

def getInput():
    points = []
    n = int(input("Enter the number of points : "))
    i=n
    while i>0:
        k,l = map(int, input("Enter Point " +str(n-i+1) + " : ").split(","))
        points.append([k,l])
        i=i-1
    
    return points

def copyArray(arr):
    arr2 = []
    for i in arr:
        arr2.append(i)
    return arr2

# function to check distance between points in a strip if they are present
def closestStrip(strip, size, d):
    min_val = d 
    for i in range(size):
        j = i + 1
        while j < size and (strip[j][1] - 
                            strip[i][1]) < min_val:
            min_val = getDistance(strip[i], strip[j])
            j += 1
  
    return min_val 

# utility function to find the distance
def closestShipUtil(points,q,n):
    if(len(points) <= 3): #use bruteforce -- base case
        return bruteForce(points,n)

    m = n//2
    midPoint = points[m]
    leftSpace = points[:m]
    rightSpace = points[m:]

    distLeft = closestShipUtil(leftSpace,q, m)
    distRight = closestShipUtil(rightSpace,q, n-m)

    minDist =  min(distLeft, distRight)

    stripP = []
    stripQ = []
    
    total = leftSpace + rightSpace

    for i in range(n):
        if(abs(total[i][0]-midPoint[0]) < minDist):
            stripP.append(total[i])
        if(abs(q[i][0] - midPoint[0]) < minDist):
            stripQ.append(q[i])

    stripP.sort(key = lambda p : p[1])
    min_a = min(minDist, closestStrip(stripP,len(stripP), minDist))
    min_b = min(minDist, closestStrip(stripQ,len(stripQ), minDist))

    return min(min_a,min_b)

# Brute Force Method Used for Base Case
def bruteForce(P, n):
    min_val = float('inf') # initialize to infinity
    for i in range(n):
        for j in range(i + 1, n):
            if getDistance(P[i], P[j]) < min_val:
                min_val = getDistance(P[i], P[j])
  
    return min_val


def closestShips(points):
    points.sort(key=lambda x: x[0]) # sort according to x-coordinates
    q = copyArray(points)
    q.sort(key=lambda x: x[1]) # sort according to y-coordinates
    n = len(points)
    return closestShipUtil(points,q,n)

points = getInput()
print("\nClosest Distance Between Two Ships is " + str(closestShips(points)))

File: Divide and Conquer Algos/Closest Pair Problem/code.cpp
#include <bits/stdc++.h>
#include <float.h>
#include <math.h>
using namespace std;

/*
    CLOSEST PAIR OF POINTS USING DIVIDE AND CONQUER ALGORITHM

    TIME COMPLEXITY: O(nlogn)
    T(n) = 2T(n/2) + O(n) + O(n)
    
    SPACE COMPLEXITY: O(logn)
*/

struct Point{
    int x;
    int y;
};

int compareX(const void *x1, const void *x2){
    Point *p1 = (Point *)x1, *p2 = (Point *)x2;
    return (p1->x!=p2->x) ? (p1->x - p2->x ): (p1->y - p2->y);
}

int compareY(const void *x1, const void *x2){
    Point *p1 = (Point *)x1, *p2 = (Point *)x2;
    return (p1->y!=p2->y) ? (p1->y - p2->y): (p1->x - p2->x);
}

float min(float x, float y){
    return x>y ? y : x;
}

float getDistance(Point *p1,Point *p2){
    float d1 = p2->x - p1->x;
    float d2 = p2->y - p1->y;
    return sqrt(d1*d1 + d2*d2);
}

float bruteForce(Point p[], int n){
    float min = FLT_MAX;
    for(int i=0;i<n;i++){
        for(int j=i+1;j<n;j++){
            if(getDistance(&p[i],&p[j]) < min){
                min = getDistance(&p[i], &p[j]);
            }
        }
    }

    return min;
}

float stripClosest(Point strip[],int size, float d){
    float min = d;
    for(int i=0;i<size;i++){
        for(int j=i+1;j<size && (strip[j].y-strip[i].y)<min;++j){
            if(getDistance(&strip[i],&strip[j])< min){
                min = getDistance(&strip[i], &strip[j]);
            }
        }
    }

    return min;
}

float closestUtil(Point Px[], Point Py[], int n){
    if(n <= 3){
        return bruteForce(Px, n);
    }

    int mid = n/2;
    Point midPoint = Px[mid];

    Point Pyl[mid];
    Point Pyr[n-mid];
    int li=0,ri=0;
    for(int i=0;i<n;i++){
        if((Py[i].x < midPoint.x || (Py[i].x == midPoint.x && Py[i].y < midPoint.y)) && li<mid)
            Pyl[li++] = Py[i];
        else    
            Pyr[ri++] = Py[i];
    }

    float dl = closestUtil(Px, Pyl,mid);
    float dr = closestUtil(Px+mid, Pyr, n-mid);

    float d = min(dl,dr);

    Point strip[n];
    int j=0;
    for(int i=0;i<n;i++){
        if(abs(Py[i].x - midPoint.x) < d){
            strip[j] = Py[i],j++;
        }
    }

    return stripClosest(strip,j,d);
}

float closest(Point p[], int n){
    Point Px[n];
    Point Py[n];

    for(int i=0;i<n;i++){
        Px[i] = p[i];
        Py[i] = p[i];
    }

    // sort according to x coordinates
    qsort(Px, n, sizeof(Point), compareX);
    // sort according to y coordinates
    qsort(Py, n, sizeof(Point), compareY);

    return closestUtil(Px, Py, n);
}

int main(){
    // Point p1 = {10,11};
    // Point p2 = {5,10};
    // cout << getDistance(&p1, &p2) << endl;

    Point P[] = {{2, 3}, {12, 30}, {40, 50}, {5, 1}, {12, 10}, {3, 4}};
    int n = sizeof(P) / sizeof(P[0]);
    cout << closest(P,n) << endl;

    return 0;
}